### PR TITLE
gh-102778: Add sys.last_exc, deprecate sys.last_type, sys.last_value,sys.last_traceback

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-102778

### CODE REVIEW

File: Doc/library/sys.rst [hunk 1]
```diff
@@ -1102,22 +1102,25 @@ always available.
...
-   Their intended use is to allow an interactive user to import a debugger module
-   and engage in post-mortem debugging without having to re-execute the command
-   that caused the error.  (Typical use is ``import pdb; pdb.pm()`` to enter the
-   post-mortem debugger; see :mod:`pdb` module for
-   more information.)
-
-   The meaning of the variables is the same as that of the return values from
-   :func:`exc_info` above.
-
+   These three variables are deprecated, use ``sys.last_exc`` instead.
```
Comment (AlexWaygood) [inline Doc/library/sys.rst]: (I know `sys.last_exc` is just immediately above this entry, but that won't necessarily be obvious to people who've arrived on the description of `sys.last_type` by clicking a link that takes them straight here, so I think linking to `sys.last_exc` is still useful here)

```suggestion
   These three variables are deprecated; use :data:`sys.last_exc` instead.
```

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -397,6 +397,12 @@ sys
   with contributions from Gregory P. Smith [Google] and Mark Shannon
   in :gh:`96123`.)
 
+* Add ``sys.last_exc`` which holds the last unhandled exception that
+  was raised (for post-mortem debugging use cases). Deprecate the
+  three fields that have the same information in its legacy form:
+  ``sys.last_type``, ``sys.last_value`` and ``sys.last_traceback``.
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* Add :data:`sys.last_exc` which holds the last unhandled exception that
  was raised (for post-mortem debugging use cases). Deprecate the
  three fields that have the same information in its legacy form:
  :data:`sys.last_type`, :data:`sys.last_value` and :data:`sys.last_traceback`.
```

File: Doc/whatsnew/3.12.rst [hunk 3]
```diff
@@ -488,6 +494,9 @@ Deprecated
   contain the creation time, which is also available in the new ``st_birthtime``
   field. (Contributed by Steve Dower in :gh:`99726`.)
 
+* The ``sys.last_type``, ``sys.last_value`` and ``sys.last_traceback`` fields.
+  Use ``sys.last_exc`` instead. (Contributed by Irit Katriel in :gh:`102778`.)
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* The :data:`sys.last_type`, :data:`sys.last_value` and :data:`sys.last_traceback`
  fields. Use :data:`sys.last_exc` instead.
  (Contributed by Irit Katriel in :gh:`102778`.)
```

File: Misc/NEWS.d/next/Core and Builtins/2023-03-17-13-43-34.gh-issue-102778.ANDv8I.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Add ``sys.last_exc`` and deprecate ``sys.last_type``, ``sys.last_value`` and
+``sys.last_traceback``, which have the same info in its legacy form.
```
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Core and Builtins/2023-03-17-13-43-34.gh-issue-102778.ANDv8I.rst]:
```suggestion
Add :data:`sys.last_exc` and deprecate :data:`sys.last_type`, :data:`sys.last_value`
and :data:`sys.last_traceback`,
which hold the same information in its legacy form.
```

**Review by AlexWaygood — COMMENTED** [review]
A few more docs nits

**Review by gvanrossum — APPROVED** [review]
Shouldn’t pdb.pm() be changed to use sys.last_exc?

**Comment by iritkatriel** [comment]
> Shouldn’t pdb.pm() be changed to use sys.last_exc?

Yes, idlelib too.  But I'm not sure whether we do that now or in 3.14 when it won't interfere with backports.

File: Doc/whatsnew/3.12.rst [hunk 5]
```diff
@@ -488,6 +494,10 @@ Deprecated
   contain the creation time, which is also available in the new ``st_birthtime``
   field. (Contributed by Steve Dower in :gh:`99726`.)
 
+* The :data:`sys.last_type`, :data:`sys.last_value` and :data:`sys.last_traceback`
+  fields. Use :data:`sys.last_exc` instead.
```
Comment (iritkatriel) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  fields are deprecated. Use :data:`sys.last_exc` instead.
```

**Comment by gvanrossum** [comment]
> > Shouldn’t pdb.pm() be changed to use sys.last_exc?
> 
> Yes, idlelib too. But I'm not sure whether we do that now or in 3.14 when it won't interfere with backports.

I would do it now, but look both (in case user code sets one or the other): if sys.last_exc exists, use that, else if sys.last_(type,value,traceback) exist, use those, else fail. Then in the future we can delete the code checking for the latter.

**Comment by terryjreedy** [comment]
Please revert the changes to idlelib.  (Why the rush before I could respond.)  I dislike the awkward workaround changes, which IDLE does not need, and they cannot be backported.  I would much prefer to make more extensive changes that get rid of nearly all references to the deprecated attributes and which *can* be backported.

For instance, in test_stackviewer setup, *replace* `svs.last_type ... = svs.exec_info()` with `svs.last_exc = svs.exec_info()[1]`.  In the teardown, replace `del ...` with `del svs.last_exc`, leaving no usage of the deprecated names.  In other modules, add `sys.last_exc = sys.last_value` and eliminate all other uses of the old names.  In June 2024, eliminate the use of last_value.

**Comment by iritkatriel** [comment]
Sorry, will revert and fix.

**Comment by iritkatriel** [comment]
> In other modules, add `sys.last_exc = sys.last_value` and eliminate all other uses of the old names.

If a user tries to access sys.last_type or sys.last_traceback then it won't work. Is this not an issue?

**Comment by terryjreedy** [comment]
Yes, `sys.last_type, sys.last_value, sys.last_traceback = excinfo` should remain as long as the REPL does the same.
I believe other settings are only for internal use in stackviewer and its test, but will check.

**Pull request merged**